nupic/research/frameworks/pytorch/imagenet/mixins/save_final_checkpoint.py

```diff
@@ -0,0 +1,134 @@
+                                    * {trainable_id} - inferred from logdir
+
+    Example
+    ```
+    config["name"] = "my_experiment"
+    config["logdir"] = ".../ImagenetTrainable_...-45xrhk_mc2"
+    config["wandb_args"] = dict(..., project="my_project")
+    config["copy_checkpoint_dir"] = "/mnt/.../{wandb_project}/{name}/{trainable_id}/"
+
+    ... 200 epochs of training
+
+    >> checkpoint saved to "/mnt/.../my_project/my_experiment/45xrhk_mc2/checkpoint_200"
+    ```
+    """
+
+    def setup_experiment(self, config):
+        """Setup experiment and get desired path for copied checkpoint."""
+        super().setup_experiment(config)
+
+        assert self.logdir
+        assert "copy_checkpoint_dir" in config
+
+        # Set default formatting.
+        self.copy_checkpoint_dir = config["copy_checkpoint_dir"]
+        self.name = config.get("name", "unknown_name")
+        self.wandb_project = "unknown_project"
+        self.trainable_id = "unknown_id"
+
+        # Retrieve project name if any - this may or may not be used.
+        if "{wandb_project}" in self.copy_checkpoint_dir:
+            wandb_args = config.get("wandb_args", {})
+            self.wandb_project = wandb_args.get("project", "unknown_project")
+
+        # Retrieve run_id - this may or may not be used.
+        if "{trainable_id}" in self.copy_checkpoint_dir:
+            self.trainable_id = self.logdir.split("-")[-1]  # e.g. 45xrhk_mc2
+
+        # Set up to directory that will contain the copied checkpoint.
+        self.setup_copy_dir()
+
+    def setup_copy_dir(self):
+        base_path = self.copy_checkpoint_dir.split("{")[0]
+        assert os.path.exists(base_path), (
+            "Can't copy to directory ({base_path}) which does not exists.")
+
+        copy_ckpt_path = self.copy_checkpoint_dir.replace("{name}", self.name)
+        copy_ckpt_path = copy_ckpt_path.replace("{wandb_project}", self.wandb_project)
+        copy_ckpt_path = copy_ckpt_path.replace("{trainable_id}", self.trainable_id)
+        self.copy_ckpt_path = Path(copy_ckpt_path)
+
+        # Create the save directory and parents as needed.
+        self.logger.info(f"Making dir for final checkpoint: {str(self.copy_ckpt_path)}")
+        self.copy_ckpt_path.mkdir(exist_ok=True, parents=True)
+
+    def stop_experiment(self):
+        """
+        Copy over last checkpoint to specified directory.
+        """
+        super().stop_experiment()
+        if self.rank != 0:
+            return
+
+        # Find the latest checkpoint.
+        log_path = Path(self.logdir)
+        checkpoint_paths = []
+        for path in log_path.iterdir():
+            if "checkpoint" in path.name:
+                num = int(path.name.split("_")[1])
+                checkpoint_paths.append((num, path))
+
+        checkpoint_paths = sorted(checkpoint_paths, key=itemgetter(0))
+        latest_checkpoint = checkpoint_paths[-1][1]  # path of latest
```

Comment:
See [nupic.research.support.ray_utils.get_last_checkpoint](https://github.com/numenta/nupic.research/blob/master/nupic/research/support/ray_utils.py#L114)

Reply:
This is not a ray specific mixin, however. I wanted something that could generalize to some directory of checkpoint (not necessarily including an `experiment_state` json file.

I've added a comment to the docstring that specifies the supported checkpoint-name format, and I've added an additional logic in case nothing is found.